lib/aptible/cli/agent.rb

```diff
@@ -104,14 +117,38 @@ def login
           token_options[:expires_in] = duration
           token = Aptible::Auth::Token.create(token_options)
         rescue OAuth2::Error => e
-          if e.code == 'otp_token_required'
-            token_options[:otp_token] = options[:otp_token] ||
-                                        ask('2FA Token: ')
-            retry
+          if e.code != 'otp_token_required'
+            raise Thor::Error, 'Could not authenticate with given ' \
+                               "credentials: #{e.code}"
+          end
+
+          u2f = (e.response.parsed['exception_context'] || {})['u2f']
+
+          if u2f && security_key
+            origin = Aptible::Auth::Resource.new.get.href
+            app_id = Aptible::Auth::Resource.new.utf_trusted_facets.href
+
+            challenge = u2f.fetch('challenge')
```

Comment:
We have to have the U2F challenge before we can authenticate with the security key and right now we get it from the `OAuth2::Error` thrown by `Aptible::Auth::Token.create`. If there's a way to get the challenge outside of that error, this might be possible but I don't think it's worth the effort at this time.

Reply:
Yeah good point. I think my original comment wasn't super clear either. As it stands, the security key doesn't really impact anything until the `rescue` block and it's not immediately clear to a reader that what occurs inside the rescue is considered "normal" in that case.

I sort of had in mind one of two things

1) Just ditch the new flag entirely. If the API prompts for U2F then ask for it if the `u2f-host` is there, otherwise fallback to OTP. This resolves the need for a user to learn about and consistently employ a flag to get this feature. The comment clarifies that this is an expected code path in normal operation.

```diff
diff --git a/lib/aptible/cli/agent.rb b/lib/aptible/cli/agent.rb
index 970f846..7bc4195 100644
--- a/lib/aptible/cli/agent.rb
+++ b/lib/aptible/cli/agent.rb
@@ -84,16 +84,7 @@ module Aptible
       option :lifetime, desc: 'The duration the token should be valid for ' \
                               '(example usage: 24h, 1d, 600s, etc.)'
       option :otp_token, desc: 'A token generated by your second-factor app'
-      option :security_key, desc: 'Use Security Key authentication if available'
       def login
-        security_key = options[:security_key]
-
-        if security_key && which('u2f-host').nil?
-          puts 'u2f-host is not installed. --security-key option cannot be' \
-               'used.'
-          return
-        end
-
         email = options[:email] || ask_then_line('Email: ')
         password = options[:password] || ask_then_line(
           'Password: ', echo: false
@@ -116,6 +107,9 @@ module Aptible
 
           token_options[:expires_in] = duration
           token = Aptible::Auth::Token.create(token_options)
+
+        # If the user has 2FA enabled, then we expect an error back from
+        # the Auth API prompting it to be entered
         rescue OAuth2::Error => e
           if e.code != 'otp_token_required'
             raise Thor::Error, 'Could not authenticate with given ' \
@@ -124,7 +118,9 @@ module Aptible
 
           u2f = (e.response.parsed['exception_context'] || {})['u2f']
 
-          if u2f && security_key
+          # Default to u2f if it's an option as long as the host binary
+          # is available
+          if u2f && which('u2f-host')
             origin = Aptible::Auth::Resource.new.get.href
             app_id = Aptible::Auth::Resource.new.utf_trusted_facets.href
```

2. Pull out the "happy" path use of rescue to a function an invoke that in `login`. This way the `rescue` block is really stuff we don't "expect" to happen if the flag is provided. This is probably a hairier thing to do and may not be worth it. I haven't thought through it enough to feel super strongly about it.